torchmetrics/functional/image/d_lambda.py

```diff
@@ -0,0 +1,141 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Tuple
+
+import torch
+from torch import Tensor
+from typing_extensions import Literal
+
+from torchmetrics.functional.image.uqi import universal_image_quality_index
+from torchmetrics.utilities.checks import _check_same_shape
+from torchmetrics.utilities.distributed import reduce
+
+
+def _d_lambda_update(ms: Tensor, fused: Tensor) -> Tuple[Tensor, Tensor]:
+    """Updates and returns variables required to compute Spectral Distortion Index. Checks for same shape and type
+    of the input tensors.
+
+    Args:
+        ms: Low resolution multispectral image
+        fused: High resolution fused image
+    """
+
+    if ms.dtype != fused.dtype:
+        raise TypeError(
+            f"Expected `ms` and `fused` to have the same data type. Got ms: {ms.dtype} and fused: {fused.dtype}."
+        )
+    _check_same_shape(ms, fused)
+    if len(ms.shape) != 4:
+        raise ValueError(
+            f"Expected `ms` and `fused` to have BxCxHxW shape. Got ms: {ms.shape} and fused: {fused.shape}."
+        )
+    return ms, fused
+
+
+def _d_lambda_compute(
+    ms: Tensor,
+    fused: Tensor,
+    p: int = 1,
+    reduction: Literal["elementwise_mean", "sum", "none"] = "elementwise_mean",
+) -> Tensor:
+    """Computes Spectral Distortion Index.
+
+    Args:
+        ms: Low resolution multispectral image
+        fused: High resolution fused image
+        p: a parameter to emphasize large spectral difference (default: 1)
+        reduction: a method to reduce metric score over labels.
+
+            - ``'elementwise_mean'``: takes the mean (default)
+            - ``'sum'``: takes the sum
+            - ``'none'``: no reduction will be applied
+
+    Example:
+        >>> import torch
+        >>> _ = torch.manual_seed(42)
+        >>> ms = torch.rand([16, 3, 16, 16])
+        >>> fused = ms * 0.75
+        >>> ms, target = _d_lambda_update(ms, fused)
+        >>> _d_lambda_compute(ms, fused)
+        tensor(3.4769e-08)
```

Comment:
Is there any way to skip the doc tests? In `uqi` similar test was passing as we had a constant factor of 0.75 for distortion. (and its a full reference metric so it will give same result without `torch.manual_seed`). Probably `torch.manual_seed` is not solving the problem here as the test is failing everytime with a different expected value.

Reply:
@Borda, I just checked the CI, it seems to be failing for multiple environments with different values (`Got: some random value`) even after using `manual_seed`.

Eg: https://github.com/PyTorchLightning/metrics/runs/5382467199?check_suite_focus=true,  https://github.com/PyTorchLightning/metrics/runs/5382472779?check_suite_focus=true

Reply:
@ankitaS11 In the previous scenario, it seems like as `ms` and `fused` were too similar, the value of this metric was really close to 0 and there was a discrepancy on the 8th decimal place across some OS/torch versions. I change `fused = ms * 0.75` to `fused = torch.rand([16, 3, 16, 16])` so that the tensors won't be so similar. Hopefully, it looks like everything works fine now :]